elements/lisk-transaction-pool/src/min_heap.ts

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright © 2020 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { Node } from './node';
+
+export class MinHeap<T, K = bigint | number> {
+	protected _nodes: Array<Node<T, K>>;
+
+	public constructor(heap?: MinHeap<T, K>) {
+		this._nodes = [];
+		if (heap) {
+			this._insertAll(heap);
+		}
+	}
+
+	public push(key: K, value: T): void {
+		const node = new Node<T, K>(key, value);
+		this._nodes.push(node);
+		this._moveUp(this._nodes.length - 1);
+	}
+
+	public pop(): { key: K; value: T } | undefined {
+		if (this.count <= 0) {
+			return undefined;
+		}
+		if (this.count === 1) {
+			const node = this._nodes[0];
+			this.clear();
+
+			return node;
+		}
+		const rootNode = this._nodes[0];
+		this._nodes[0] = this._nodes.pop() as Node<T, K>;
+		this._moveDown(0);
+
+		return rootNode;
+	}
+
+	public peek(): { key: K; value: T } | undefined {
+		if (this._nodes.length <= 0) {
+			return undefined;
+		}
+
+		return this._nodes[0];
+	}
+
+	public clone(): MinHeap<T, K> {
+		return new MinHeap(this);
+	}
+
+	public clear(): void {
+		this._nodes = [];
+	}
+
+	public get count(): number {
+		return this._nodes.length;
+	}
+
+	public get keys(): ReadonlyArray<K> {
+		return this._nodes.map(n => n.key);
+	}
+
+	public get values(): ReadonlyArray<T> {
+		return this._nodes.map(n => n.value);
+	}
+
+	private _insertAll(heap: MinHeap<T, K>): void {
+		if (!(heap instanceof MinHeap)) {
+			throw new Error('Cannot insertAll from other than heap');
```

Comment:
Maybe a better message would be what the error checks for like for example `Can only insert heap instances`?